### ISSUE TITLE
Memory leak using merged stream in parJoinUnbounded

### ISSUE THREAD

**Issue by oleg-py** [typelevel/fs2#2598]
fs2 3.1.1 / cats-effect 3.2.5 / Scala 2.13.5

---

Minimized reproduction:

```scala
import cats.effect.{IO, IOApp}
import cats.implicits._
import scala.concurrent.duration._

object MergeParJoinLeak extends IOApp.Simple {
  case class Box(i: Int)
  val src = fs2.Stream(1).delayBy[IO](10.millis).repeat.scanMonoid.map(Box(_))
  val run = fs2.Stream(src merge fs2.Stream()).parJoinUnbounded
    .compile.drain
}
```
This app continuously leaks `Box`es and a lot of fs2 Stream internal objects:
![изображение](https://user-images.githubusercontent.com/4143670/132464909-3823fb72-c5b7-4a30-8d6b-db4b0fde984b.png)
Apparently, not a single `Box` is released. Most stream bits seem to be nested, with the selected pattern repeating:
![изображение](https://user-images.githubusercontent.com/4143670/132466482-547bcdfa-7690-428f-a6a5-9631a664d3af.png)

---

This doesn't happen if you remove _either_ merge or parJoinUnbounded. It also doesn't leak if you replace parJoinUnbounded with merge (so that's a viable workaround)

**Comment by mpilquist** [comment]
Thanks! I added a further minimization as a test here: https://github.com/typelevel/fs2/pull/2599/files

**Comment by diesalbla** [comment]
@oleg-py The leak reported, with some minimisations, is covered by the new tests added https://github.com/typelevel/fs2/pull/2602, which also seems to fix the issue.

**Comment by mpilquist** [comment]
Fixed in 3.1.2
